internal/objectstore/client.go

```diff
@@ -0,0 +1,80 @@
+package objectstore
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+)
+
+type awsS3Client interface {
+	PutObject(ctx context.Context, input *s3.PutObjectInput, opts ...func(*s3.Options)) (*s3.PutObjectOutput, error)
+	GetObject(ctx context.Context, input *s3.GetObjectInput, opts ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+type S3Client struct {
+	bucketName string
+	awsClient  awsS3Client
+}
+
+func (c *S3Client) Put(objectKey string, obj any) (*s3.PutObjectOutput, error) {
+	b, err := json.Marshal(obj)
+	if err != nil {
+		return &s3.PutObjectOutput{}, err
+	}
+
+	return c.awsClient.PutObject(
+		context.Background(),
```

Comment:
it would be good to pass the context down from the top-level here, and in the others